Implement positive-definite TVD methods for tracer transport in EDMF-TKE PBL schemes and SAMF convective schemes

rmontuoro opened the pull request:
This PR introduces positive-definite Total Variation Diminishing (TVD) mass-flux methods recently developed at NOAA/EMC (Jongil Han) to improve tracer transport in EDMF-TKE PBL schemes and SAMF cumulus convective parameterizations.

The included code changes have been tested on Hera/Intel and Orion/Intel platforms.

Please see issue #753 for details. Results of MJO and jet stream diagnostic tests are available in [this document](https://github.com/NCAR/ccpp-physics/files/7297212/Diagnostics_ufsp6_ufsp7_ufsp7tvd.pptx.pdf).

SMoorthi-emc commented:
These TVD related changes are made by Jongil Han.  Yes, it will change results.

JongilHan66 reviewed (approved):
The changes look good.

climbfuji commented:
@grantfirl @llpcarson Can you please review this PR? We are hoping to merge it next Monday. @SMoorthi-emc @yangfanglin I want to confirm with you that you are ok with the proposed changes. Thanks!

climbfuji commented:
@rmontuoro I see there is a merge conflict for `physics/satmedmfvdifq.meta`. I guess you will have to pull in the latest updates from the authoritative repositories for all your PRs (note the upcoming ufs-weather-model HAFS doc update for the top-level repository, fv3atm and ccpp-physics won't change before your PR).

yangfanglin commented:
These changes are important not only for the updated schemes themselves but
also for other physics processes like microphysics and atmospheric
compositions.  The developers have rigorously tested the impact on
sciences.  So I approve this PR if there are no code merging issues.

On Fri, Nov 5, 2021 at 10:44 AM Dom Heinzeller ***@***.***>
wrote:

> @grantfirl <https://github.com/grantfirl> @llpcarson
> <https://github.com/llpcarson> Can you please review this PR? We are
> hoping to merge it next Monday. @SMoorthi-emc
> <https://github.com/SMoorthi-emc> @yangfanglin
> <https://github.com/yangfanglin> I want to confirm with you that you are
> ok with the proposed changes. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/ccpp-physics/pull/754#issuecomment-961952759>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKY5N2KBJM2OGQ2TRFXBW6DUKP3VVANCNFSM5FPYHL5Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
*Fanglin Yang, Ph.D.*
*Chief, Model Physics Group*
*Modeling and Data Assimilation Branch*

*NOAA/NWS/NCEP Environmental Modeling Center*

*https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/
<https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/>*

SMoorthi-emc commented on physics/samfdeepcnv.f:
```diff
@@ -1179,6 +1200,28 @@ subroutine samfdeepcnv_run (im,km,itc,ntc,cliq,cp,cvap,           &
+     &                     (ctro(i,k,kk)+ctro(i,k-1,kk)))/factor
+                   chem_c(i,k,n)=fscav(n)*ecko(i,k,kk)
+                   tem=chem_c(i,k,n)/(1.+c0t(i,k)*dz)
+                   chem_pw(i,k,n)=c0t(i,k)*dz*tem*eta(i,k-1)
```
I would have put blanks arout "=".

rmontuoro replied:
Agree.

SMoorthi-emc commented on physics/samfdeepcnv.f:
```diff
@@ -2009,8 +2052,8 @@ subroutine samfdeepcnv_run (im,km,itc,ntc,cliq,cp,cvap,           &
      &                   - heo(i,1)) * grav / dp
-          dellaq(i,1) = edto(i) * etad(i,1) * (qrcdo(i,1)
-     &                   - qo(i,1)) * grav / dp
+          dellaq(i,1) = edto(i) * etad(i,1) * qrcdo(i,1)
```
If you use dpi = grav / (1000.0*del(i,1), then you could eliminate multiple divides

rmontuoro replied:
Agree.

@climbfuji - do we still have time to implement these suggestions?

climbfuji replied:
Absolutely!

SMoorthi-emc reviewed (approved):
There are lots of changes in the code and I assume Jongil has checked it well.
I have not verified everything.

rmontuoro commented:
> @rmontuoro I see there is a merge conflict for `physics/satmedmfvdifq.meta`. I guess you will have to pull in the latest updates from the authoritative repositories for all your PRs (note the upcoming ufs-weather-model HAFS doc update for the top-level repository, fv3atm and ccpp-physics won't change before your PR).

Just updated all PRs to the latest code revisions.

rmontuoro commented:
@SMoorthi-emc - I've implemented your recommendations and added a few more changes to improve performance for the deep & shallow convective schemes.

@SMoorthi-emc and @JongilHan66 - could you please take a look at the changes and let me know if you agree with them?

SMoorthi-emc commented on physics/samfdeepcnv.f:
```diff
@@ -1992,18 +1992,18 @@ subroutine samfdeepcnv_run (im,km,itc,ntc,cliq,cp,cvap,           &
+              dt = to(i,k)
+              dg = gamma
+              dh = heso(i,k)
+              dz = -1.*(zo(i,k+1)-zo(i,k))
```
Not trying to be picky, but we really don't need this multiplication by "-1".

SMoorthi-emc replied:
it could be just a "-" or you could reverse the two items inside "(  )"

rmontuoro replied:
Agree!

rmontuoro replied:
@SMoorthi-emc - I've implemented your suggested changes.

SMoorthi-emc commented:
great.

On Mon, Nov 8, 2021 at 2:28 PM Raffaele Montuoro ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In physics/samfdeepcnv.f
> <https://github.com/NCAR/ccpp-physics/pull/754#discussion_r745029107>:
>
> > @@ -1992,18 +1992,18 @@ subroutine samfdeepcnv_run (im,km,itc,ntc,cliq,cp,cvap,           &
>          do i = 1, im
>            if (cnvflg(i) .and. k < jmin(i)) then
>                gamma = el2orc * qeso(i,k) / to(i,k)**2
> -              dhh=hcdo(i,k)
> -              dt=to(i,k)
> -              dg=gamma
> -              dh=heso(i,k)
> -              dz=-1.*(zo(i,k+1)-zo(i,k))
> +              dhh = hcdo(i,k)
> +              dt = to(i,k)
> +              dg = gamma
> +              dh = heso(i,k)
> +              dz = -1.*(zo(i,k+1)-zo(i,k))
>
> @SMoorthi-emc <https://github.com/SMoorthi-emc> - I've implemented your
> suggested changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/ccpp-physics/pull/754#discussion_r745029107>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALLVRYQVI4E3EMKJKHTAMDLULAQFVANCNFSM5FPYHL5Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Dr. Shrinivas Moorthi
Research Meteorologist
Modeling and Data Assimilation Branch
Environmental Modeling Center / National Centers for Environmental
Prediction
5830 University Research Court - (W/NP23), College Park MD 20740 USA
Tel: (301)683-3718

e-mail: ***@***.***
Phone:  (301) 683-3718   Fax: (301) 683-3718

junwang-noaa commented:
@JongilHan66 Are the code changes looking good to you?

JongilHan66 commented:
The changes which I agree on look good and would save some computing time.
